### ISSUE TITLE
Possible error in Ch12 mlp code

### ISSUE THREAD

**Issue by rvashurin** [rasbt/python-machine-learning-book-3rd-edition#154]
This might be just my misunderstanding, but it might also be an error.

The code of an mlp provided in the chapter computes two sigmoid activations - one for the hidden layer, and one for the output layer (see https://github.com/rasbt/python-machine-learning-book-3rd-edition/blob/master/ch12/neuralnet.py#L78 and https://github.com/rasbt/python-machine-learning-book-3rd-edition/blob/master/ch12/neuralnet.py#L87). But when we backpropagate the error, it seems that we discard one of the sigmoid gradients - namely the one for output layer. The gradient of loss function w.r.t. output weight matrix `grad_w_out` is computed without using derivative of sigmoid activation: https://github.com/rasbt/python-machine-learning-book-3rd-edition/blob/master/ch12/neuralnet.py#L87. This formula would be clear for me if it had been multiplied by sigmoid derivative `a_out * (1 - a_out)`, but in it's current form it just doesn't click for me.

Same goes for the computation of `grad_w_h` which only considers activation function of hidden layer, but doesn't seem to rely on output activation as well.

Could you elaborate on this?

**Comment by rasbt** [comment]
Hi there,
I am currently traveling and can't check it out in detail  at the moment, but it reminds me of a previous question with an answer that might answer your question: https://github.com/rasbt/python-machine-learning-book-2nd-edition/issues/54

I agree that in the chapter, the notation and implementation are not clear, I should clarify this better in future editions.

**Comment by rvashurin** [comment]
Thanks for the quick reply!

You are right, didn't make the connection myself. But I do think that the chapter could use a bit more detailed math explanation. I used this paper to figure things out eventually, I think it's pretty clear: http://machinelearningmechanic.com/assets/pdfs/cross_entropy_loss_derivative.pdf, the author directly computed the derivative of the loss with regards to combined input to the softmax layer, not it's activations.
